### PR TITLE
[RISCV] Add integer RISCVISD::SELECT_CC to canCreateUndefOrPoison and isGuaranteedNotToBeUndefOrPoison.

### PR DESCRIPTION
**Pull request by topperc**
Integer RISCVISD::SELECT_CC doesn't create poison. If none of the, operands are poison, the result is not poison.

This allows ISD::FREEZE to be hoisted above RISCVISD::SELECT_CC.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -17041,6 +17041,38 @@ unsigned RISCVTargetLowering::ComputeNumSignBitsForTargetNode(
...
+bool RISCVTargetLowering::isGuaranteedNotToBeUndefOrPoisonForTargetNode(
+    SDValue Op, const APInt &DemandedElts, const SelectionDAG &DAG,
+    bool PoisonOnly, unsigned Depth) const {
+
+  // TODO: Add more target nodes.
+  switch (Op.getOpcode()) {
+  case RISCVISD::SELECT_CC:
+    return all_of(Op->ops(), [&](SDValue V) {
+      return DAG.isGuaranteedNotToBeUndefOrPoison(V, PoisonOnly, Depth + 1);
+    });
```
Comment (nikic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:17053]: This implementation doesn't seem to be consistent with canCreateUndefOrPoisonForTargetNode() in terms of FP predicate handling?

Also @RKSimon why https://github.com/llvm/llvm-project/blob/a84eb244129f288d609307ad42ab5e6c8e1cc795/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L5035-L5038 ? I would expect that in a case like this implementing canCreateUndefOrPoisonForTargetNode is sufficient...

Comment (nikic) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:17053]: I think we should at least make that part of the default implementation of this hook.

Comment (RKSimon) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:17053]: Sorry, I can't remember off hand why it ended up that way - I suppose it was mainly to make it so we could return directly from isGuaranteedNotToBeUndefOrPoisonForTargetNode? Theres sufficiently little target support so far that refactoring isn't out of the question if it proves necessary.

**Comment by topperc** [comment]
Ping

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:17055]: I'd still like to see this change. It makes very little sense to spell this out per-target. The code in https://github.com/llvm/llvm-project/blob/a84eb244129f288d609307ad42ab5e6c8e1cc795/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L5039-L5043 should be in the default impl of isGuaranteedNotToBeUndefOrPoisonForTargetNode, and then you won't have to do anything special here.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -3786,7 +3786,12 @@ bool TargetLowering::isGuaranteedNotToBeUndefOrPoisonForTargetNode(
        Op.getOpcode() == ISD::INTRINSIC_VOID) &&
       "Should use isGuaranteedNotToBeUndefOrPoison if you don't know whether Op"
       " is a target node!");
-  return false;
+
+  return !canCreateUndefOrPoisonForTargetNode(Op, DemandedElts, DAG, PoisonOnly,
+                                              /*ConsiderFlags*/ true, Depth) &&
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:3791]:
```suggestion
                                              /*ConsiderFlags=*/ true, Depth) &&
```

**Review by nikic — APPROVED** [review]
LGTM

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**